src/core/mesh/qgstriangularmesh.cpp

```diff
@@ -58,19 +58,27 @@ static void ENP_centroid( const QPolygonF &pX, double &cx, double &cy )
 
   double signedArea = 0.0;
 
+  const QPointF &pt0 = pX.first();
+  QPolygonF localPolygon( pX.count() );
+  for ( int i = 0; i < pX.count(); ++i )
+    localPolygon[i] = pX.at( i ) - pt0;
+
   // For all vertices except last
   int i = 0;
   for ( ; i < pX.size() - 1; ++i )
```

Comment:
shouldn't you loop over localPolygon instead of pX?

Reply:
> does we have any tests for this piece of code?

centroids were not directly tested, test added now.